lib/segment/src/index/hnsw_index/graph_links.rs

```diff
@@ -152,108 +167,107 @@ impl GraphLinksConverter {
+                    offsets.push((links.len() as u64) / size_of::<PointOffsetType>() as u64);
+                }
+            });
+        }
+
+        let offsets_padding = {
+            let len = links.len() + reindex.as_bytes().len();
+            len.next_multiple_of(size_of::<u64>()) - len
+        };
+
         Self {
-            edges,
+            compressed,
+            m,
+            m0,
+            links,
+            offsets,
             reindex,
-            back_index,
-            total_links_len,
-            total_offsets_len: total_offsets_len as usize,
             path: None,
             level_offsets,
-            point_count_by_level,
+            offsets_padding,
         }
     }
 
     /// Size of compacted graph in bytes.
-    pub fn data_size(&self) -> usize {
+    fn data_size(&self) -> usize {
         HEADER_SIZE
-            + self.point_count_by_level.len() * size_of::<u64>()
-            + self.reindex.len() * size_of::<PointOffsetType>()
-            + self.total_links_len * size_of::<PointOffsetType>()
-            + self.offsets_padding()
-            + self.total_offsets_len * size_of::<u64>()
-    }
-
-    fn offsets_padding(&self) -> usize {
-        (self.total_links_len + self.reindex.len()) % 2 * size_of::<u32>()
+            + self.level_offsets.as_bytes().len()
+            + self.reindex.as_bytes().len()
+            + self.links.len()
+            + self.offsets_padding
+            + self.offsets.as_bytes().len()
     }
 
     fn serialize_to_vec(&self) -> Vec<u8> {
-        let size = self.data_size();
-        let mut data = Vec::with_capacity(size);
+        let mut data = Vec::with_capacity(self.data_size());
         // Unwrap should be the safe as `impl Write` for `Vec` never fails.
         self.serialize_to_writer(&mut data).unwrap();
-        debug_assert_eq!(data.len(), size);
+        debug_assert_eq!(data.len(), self.data_size());
         data
     }
 
     fn serialize_to_writer(&self, writer: &mut impl Write) -> std::io::Result<()> {
         let header = GraphLinksFileHeader {
             point_count: self.reindex.len() as u64,
-            levels_count: self.point_count_by_level.len() as u64,
-            total_links_len: self.total_links_len as u64,
-            total_offsets_len: self.total_offsets_len as u64,
-            offsets_padding: self.offsets_padding() as u64,
+            levels_count: self.level_offsets.len() as u64,
+            total_links_len: if self.compressed {
+                self.links.len() as u64
+            } else {
+                self.links.len() as u64 / size_of::<PointOffsetType>() as u64
+            },
+            total_offsets_len: self.offsets.len() as u64,
+            offsets_padding: if self.compressed {
+                u64::MAX
+            } else {
+                self.offsets_padding as u64
+            },
+            m: if self.compressed { self.m as u16 } else { 0 },
```

Comment:
I mentioned in another thread that this behaviour is very untrivial and in the future develops may struggle of the situation where `m` and `m0` are zeroes. I proposed a couple of solutions

Reply:
Updated: these values are present in `HeaderCompressed` but not in `HeaderPlain`.